File: flutter/lib/models/order.dart
import 'package:siargao_trading_road/models/product.dart';
import 'package:siargao_trading_road/models/user.dart';
import 'package:siargao_trading_road/models/rating.dart';

class OrderItem {
  final int id;
  final int orderId;
  final int productId;
  final Product product;
  final int quantity;
  final double unitPrice;
  final double subtotal;
  final DateTime createdAt;
  final DateTime updatedAt;

  OrderItem({
    required this.id,
    required this.orderId,
    required this.productId,
    required this.product,
    required this.quantity,
    required this.unitPrice,
    required this.subtotal,
    required this.createdAt,
    required this.updatedAt,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'] as int,
      orderId: json['order_id'] as int,
      productId: json['product_id'] as int,
      product: Product.fromJson(json['product'] as Map<String, dynamic>),
      quantity: json['quantity'] as int,
      unitPrice: (json['unit_price'] as num).toDouble(),
      subtotal: (json['subtotal'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class Order {
  final int id;
  final int storeId;
  final User? store;
  final int supplierId;
  final User? supplier;
  final String status;
  final double totalAmount;
  final String? paymentMethod;
  final String? paymentStatus;
  final String? paymentProofUrl;
  final String? deliveryOption;
  final double? deliveryFee;
  final double? distance;
  final String? shippingAddress;
  final String? notes;
  final List<OrderItem> orderItems;
  final List<OrderRating>? ratings;
  final DateTime createdAt;
  final DateTime updatedAt;

  Order({
    required this.id,
    required this.storeId,
    this.store,
    required this.supplierId,
    this.supplier,
    required this.status,
    required this.totalAmount,
    this.paymentMethod,
    this.paymentStatus,
    this.paymentProofUrl,
    this.deliveryOption,
    this.deliveryFee,
    this.distance,
    this.shippingAddress,
    this.notes,
    required this.orderItems,
    this.ratings,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as int,
      storeId: json['store_id'] as int,
      store: json['store'] != null ? User.fromJson(json['store'] as Map<String, dynamic>) : null,
      supplierId: json['supplier_id'] as int,
      supplier: json['supplier'] != null ? User.fromJson(json['supplier'] as Map<String, dynamic>) : null,
      status: json['status'] as String,
      totalAmount: (json['total_amount'] as num).toDouble(),
      paymentMethod: json['payment_method'] as String?,
      paymentStatus: json['payment_status'] as String?,
      paymentProofUrl: json['payment_proof_url'] as String?,
      deliveryOption: json['delivery_option'] as String?,
      deliveryFee: json['delivery_fee'] != null ? (json['delivery_fee'] as num).toDouble() : null,
      distance: json['distance'] != null ? (json['distance'] as num).toDouble() : null,
      shippingAddress: json['shipping_address'] as String?,
      notes: json['notes'] as String?,
      orderItems: (json['order_items'] as List<dynamic>?)
          ?.map((item) => OrderItem.fromJson(item as Map<String, dynamic>))
          .toList() ?? [],
      ratings: json['ratings'] != null
          ? (json['ratings'] as List<dynamic>)
              .map((rating) => OrderRating.fromJson(rating as Map<String, dynamic>))
              .toList()
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class Message {
  final int id;
  final int orderId;
  final int senderId;
  final MessageSender sender;
  final String content;
  final String? imageUrl;
  final DateTime createdAt;

  Message({
    required this.id,
    required this.orderId,
    required this.senderId,
    required this.sender,
    required this.content,
    this.imageUrl,
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as int,
      orderId: json['order_id'] as int,
      senderId: json['sender_id'] as int,
      sender: MessageSender.fromJson(json['sender'] as Map<String, dynamic>),
      content: json['content'] as String,
      imageUrl: json['image_url'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class MessageSender {
  final int id;
  final String name;
  final String role;

  MessageSender({
    required this.id,
    required this.name,
    required this.role,
  });

  factory MessageSender.fromJson(Map<String, dynamic> json) {
    return MessageSender(
      id: json['id'] as int,
      name: json['name'] as String,
      role: json['role'] as String,
    );
  }
}

File: golang/models/order.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type OrderStatus string

const (
	OrderStatusDraft     OrderStatus = "draft"
	OrderStatusPreparing OrderStatus = "preparing"
	OrderStatusInTransit OrderStatus = "in_transit"
	OrderStatusDelivered OrderStatus = "delivered"
	OrderStatusCancelled OrderStatus = "cancelled"
)

type PaymentMethod string

const (
	PaymentMethodCashOnDelivery PaymentMethod = "cash_on_delivery"
	PaymentMethodGCash          PaymentMethod = "gcash"
)

type DeliveryOption string

const (
	DeliveryOptionPickup  DeliveryOption = "pickup"
	DeliveryOptionDeliver DeliveryOption = "deliver"
)

type PaymentStatus string

const (
	PaymentStatusPending PaymentStatus = "pending"
	PaymentStatusPaid    PaymentStatus = "paid"
	PaymentStatusFailed  PaymentStatus = "failed"
)

type Order struct {
	ID              uint           `gorm:"primaryKey" json:"id"`
	StoreID         uint           `gorm:"not null;index" json:"store_id"`
	Store           User           `gorm:"foreignKey:StoreID;references:ID" json:"store"`
	SupplierID      uint           `gorm:"not null;index" json:"supplier_id"`
	Supplier        User           `gorm:"foreignKey:SupplierID;references:ID" json:"supplier,omitempty"`
	Status          OrderStatus    `gorm:"type:varchar(20);not null;default:'draft'" json:"status"`
	TotalAmount     float64        `gorm:"type:decimal(10,2);default:0" json:"total_amount"`
	PaymentMethod   PaymentMethod  `gorm:"type:varchar(20)" json:"payment_method"`
	PaymentStatus   PaymentStatus  `gorm:"type:varchar(20);default:'pending'" json:"payment_status"`
	PaymentProofURL string         `gorm:"type:varchar(500)" json:"payment_proof_url,omitempty"`
	DeliveryOption  DeliveryOption `gorm:"type:varchar(20)" json:"delivery_option"`
	DeliveryFee     float64        `gorm:"type:decimal(10,2);default:0" json:"delivery_fee"`
	Distance        float64        `gorm:"type:decimal(10,2);default:0" json:"distance"`
	ShippingAddress string         `gorm:"type:text" json:"shipping_address"`
	Notes           string         `gorm:"type:text" json:"notes"`
	OrderItems      []OrderItem    `gorm:"foreignKey:OrderID" json:"order_items"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
}

type OrderItem struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	OrderID   uint      `gorm:"not null;index" json:"order_id"`
	Order     Order     `gorm:"foreignKey:OrderID" json:"-"`
	ProductID uint      `gorm:"not null;index" json:"product_id"`
	Product   Product   `gorm:"foreignKey:ProductID" json:"product"`
	Quantity  int       `gorm:"not null" json:"quantity"`
	UnitPrice float64   `gorm:"type:decimal(10,2);not null" json:"unit_price"`
	Subtotal  float64   `gorm:"type:decimal(10,2);not null" json:"subtotal"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}
